docs/rfc/0004_component_groups_scale_view.md

```diff
@@ -0,0 +1,300 @@
+```
+
+### Priority and Replica Constraints
+
+When updating the scale subresource, it's updating the `.spec.replicas` field above. The mutating webhook will then work
+to distribute the replicas of `.spec.replicas` to the replicas under the `scalePolicy`, following the rules below:
+
+- The replicas in each group must be confined within the constraints.
+- The groups with higher priorities always consume replicas first until it reaches the max boundary. E.g.,
+  when `.spec.replicas` is 4, the 'enduring' group (with priority 1, larger than the 'spot' group's) will have 2 (which
+  is the max) replicas, and then the left is assigned to the 'spot' group.
+- The groups with the same priorities tend to have similar replicas, within the constraints. Which is to say, the target
+  is to minimize the variance.
+
+## `RisingWave` CRD and Webhooks
+
+```yaml
+apiVersion: risingwave.risingwavelabs.com/v1alpha1
+kind: RisingWave
+metadata:
+  name: hello
+# ....
+status:
+  scaleViews: # An array of scale view summaries. Maintained by the RisingWaveScaleView controller.
+  - name: hello-frontend-scale-view # Name of the RisingWaveScaleView object.
+    uid:                            # UID of the object.
+    component: frontend             # Component.
+    generation: 10                  # Generation of the RisingWaveScaleView object. It indicates the locked version.
+    groupReplicas: # Group and their desired replicas. Once a group is owned by some scale view, then another scale view
+    # would not be able to grab the lock again. It's achieved with the atomic update.
+    # If a group and its desired replicas is listed below, then any changes to the target group's replicas will be rejected
+    # unless it equals to the replicas here (hence locks that group). This should be guaranteed by the webhook. 
+    - name: 'enduring'
+      replicas: 1
+    - name: 'spot'
+      replicas: 3
+```
+
+Validating webhook:
+
+- If the webhook observes scaleViews field, it must check the replicas field of each target groups and reject any change
+  not matches.
+
+This is a compatible change and should not affect the RisingWave's workflow.
+
+## `RisingWaveScaleView` Webhooks
+
+![RisingWaveScaleView Webhooks](./images/scale_view_webhooks.png)
+
+## `RisingWaveScaleView` Controller
+
+![RisingWaveScaleView Controller](./images/scale_view_controller_workflow.png)
+
+## Use Cases
+
+Case 1. Multi-AZ frontend and balanced scale
+
+```yaml
+apiVersion: risingwave.singularity-data.com/v1alpha1
+kind: RisingWave
+metadata:
+  name: multi-az
+spec:
+  components:
+    frontend:
+      groups:
+      - name: az-1
+        replicas: 1
+      - name: az-2
+        replicas: 1
+  # ....
+---
+apiVersion: risingwave.risingwavelabs.com/v1alpha1
+kind: RisingWaveScaleView
+metadata:
+  name: multi-az-frontend-sv
+spec:
+  targetRef:
+    name: multi-az
+    component: frontend
+  scalePolicy:
+  - group: az-1
```

Comment:
Are the `group`s mapped to node groups through node-selector?

Reply:
Nope, they are directly mapped to the group in RisingWave. For the node-selectors, the relationship would be like

`group in RisingWaveScaleView` -> `group in RisingWave` -> `node selectors in that group's template`